.github/workflows/build.yml

```diff
@@ -9,49 +9,97 @@ jobs:
+          restore-keys: |
+            ${{ runner.os }}-cache-npm-
+    
       - name: Install Dependencies
         run: |
           npm install
+
       - name: Create .env
         run: |
           cat > .env << EOF
           MYANIMELIST_CLIENT_ID=${{ vars.MYANIMELIST_CLIENT_ID }}
           ANILIST_CLIENT_ID=${{ vars.ANILIST_CLIENT_ID }}
+          GIT_HASH=$(git rev-parse --short HEAD)
+          RELEASE_DATE=$(date --utc +'%d/%m/%y %I:%M %p %Z')
+          BUILD_TYPE=Github Action
           EOF
+
       - name: Cache Gradle Wrapper
-        uses: actions/cache@v3
+        uses: actions/cache@v4
         with:
           path: ~/.gradle/wrapper
           key: ${{ runner.os }}-gradle-wrapper-${{ hashFiles('gradle/wrapper/gradle-wrapper.properties') }}
 
       - name: Cache Gradle Dependencies
-        uses: actions/cache@v3
+        uses: actions/cache@v4
         with:
           path: ~/.gradle/caches
           key: ${{ runner.os }}-gradle-caches-${{ hashFiles('gradle/wrapper/gradle-wrapper.properties') }}
           restore-keys: |
             ${{ runner.os }}-gradle-caches-
 
+      - name: Cache build intermediates
+        uses: actions/cache@v4
+        with:
+          path: |
+            android/app/build/generated
+            android/app/build/intermediates
+            android/app/build/tmp
+            android/.gradle
+          key: ${{ runner.os }}-build-${{ hashFiles('package-lock.json') }}
+
       - name: Make Gradlew Executable
         run: cd android && chmod +x ./gradlew
 
       - name: Build Android Release
         run: |
           cd android && ./gradlew assembleRelease --no-daemon
+          mv app/build/outputs/apk/release/app-release.apk app/build/outputs/apk/release/LNReader.apk
 
       - name: Set Env
         run: |
           set -x
           echo "COMMIT_COUNT=$(git rev-list --count HEAD)" >> $GITHUB_ENV
+          echo "COMMIT_ID=$(git rev-parse --short HEAD)" >> $GITHUB_ENV
 
       - name: Upload Artifact
-        uses: actions/upload-artifact@v1
+        env:
+          COMMIT_COUNT: ${{ env.COMMIT_COUNT }}
+        uses: actions/upload-artifact@v4
         with:
           name: LNReader-r${{ env.COMMIT_COUNT }}
-          path: android/app/build/outputs/apk/release/app-release.apk
+          path: android/app/build/outputs/apk/release/LNReader.apk
+
+      - name: Build Custom APK
+        env:
+          COMMIT_COUNT: ${{ env.COMMIT_COUNT }}
+          COMMIT_ID: ${{ env.COMMIT_ID }}
+        run: |
+          sed -i 's/LNReader/LNReader-r${{ env.COMMIT_COUNT }}(${{ env.COMMIT_ID }})/g' android/app/src/main/res/values/strings.xml
+          cd android && ./gradlew assembleRelease --no-daemon -PcustomAppId=com.rajarsheechatterjee.LNReader.commit_${{ env.COMMIT_ID }}
+          mv app/build/outputs/apk/release/app-release.apk app/build/outputs/apk/release/LNReader-r${{ env.COMMIT_COUNT }}\(${{ env.COMMIT_ID }}\)-customID.apk
```

Comment:
Why we should build app 2times.

Reply:
For custom name and package id

Reply:
I think only custom one is enough. no need the release one. 
We dont use github action artifact for release one anyway

Reply:
When redirecting people to download the latest build it would be better if it didn't install a new app on there phones, so both are useful or only the non-custom

Reply:
We should not suggest people to override the stable build with dev build.  Dev build is more like for `testing` than `quick fix` imo

Reply:
well they have the choice

Reply:
Also, we have to consider free github actions 2000 minutes/month.
This will double execution time. it is wasting and abusing

Reply:
first build is 9min second is 1min, and i've increased the number of cached things

Reply:
i lied it's 9min when cache miss and 5min on cache hit the second one is always 1min tho